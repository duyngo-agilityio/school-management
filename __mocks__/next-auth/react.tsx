import { createContext } from "react";

export const useSession = () => ({
  data: {
    user: {
      token: "access-token",
      email: "user@gmail.com",
      id: "1",
      name: "User",
    },
    expires: Date.now().toString(),
  },
  status: "authenticated",
  update: async () => null,
});

export type Session = {
  data: {
    user: {
      token: string;
      email: string;
      id: string;
      name: string;
    };
    expires: string;
  };
  status: "loading" | "authenticated" | "unauthenticated";
  update: (data?: Session) => Promise<Session | null>;
};

const INITIAL_SESSION: Session = {
  data: {
    user: {
      token: "access-token",
      email: "user@gmail.com",
      id: "1",
      name: "User",
    },
    expires: Date.now().toString(),
  },
  status: "authenticated",
  update: async () => null,
};

const SessionContext = createContext(INITIAL_SESSION);

export const SessionProvider = ({ children }: React.PropsWithChildren) => {
  <SessionContext.Provider value={INITIAL_SESSION}>
    {children}
  </SessionContext.Provider>;
};
