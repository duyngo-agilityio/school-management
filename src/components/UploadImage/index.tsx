'use client';

import { useRef, useState } from 'react';
import { FieldError } from 'react-hook-form';

// Icons
import { TrashIcon, UploadIcon } from '@/icons';

// Components
import {
  Box,
  Flex,
  IconButton,
  Input,
  Spinner,
  Text,
  useToast,
} from '@chakra-ui/react';
import CustomImage from '../CustomImage';

// Hooks
import { useUpload } from '@/hooks';

// Constants
import { BLUR_DATA_URL } from '@/constants';

export interface IUploadImageProps {
  alt: string;
  title?: string;
  height?: string;
  width?: string;
  boxHeightImage?: string;
  src?: string;
  onChange?: (url: string) => void;
  onRemove?: () => void;
  error?: FieldError;
}

const UploadImage = ({
  title,
  height,
  width,
  alt,
  boxHeightImage,
  src,
  onChange,
  onRemove,
  error,
}: IUploadImageProps) => {
  const [previewImage, setPreviewImage] = useState(src);

  const toast = useToast();

  const { handleUploadImage, isLoading } = useUpload();

  const inputRef = useRef<HTMLInputElement>(null);

  // Handle change upload image
  const handleChange = async (e: React.ChangeEvent<HTMLInputElement>) => {
    const file = e.target.files?.[0];

    if (file) {
      const dataImage = await handleUploadImage(file);

      if (dataImage.error) {
        toast({
          title: 'Upload image fail',
          status: 'error',
        });
        return;
      }

      return { url: dataImage.image };
    }
  };

  const handleRemoveImage = () => {
    onRemove?.();
    setPreviewImage('');
  };

  const handleButtonClick = () => {
    inputRef.current?.click();
  };

  const handleUploadFile = async (e: React.ChangeEvent<HTMLInputElement>) => {
    const image = await handleChange(e);

    if (image) {
      const { url } = image;
      onChange?.(url);
      setPreviewImage(url);
    }
  };

  return (
    <>
      {!previewImage ? (
        <Flex
          w={width}
          h={height}
          flexDirection="column"
          border="1px solid"
          borderColor={error ? 'error.400' : 'borderCatSkill'}
          borderRadius="md"
          alignItems="center"
        >
          {isLoading ? (
            <Flex justifyContent="center" alignItems="center" h="inherit">
              <Spinner label="Uploading..." />
            </Flex>
          ) : (
            <Flex
              h="full"
              flexDirection="column"
              alignItems="center"
              justifyContent="center"
              onClick={handleButtonClick}
            >
              <Flex alignItems="center" cursor="pointer">
                <UploadIcon />
                <Text variant="primary" size="md" color="primaryColor">
                  {title}
                </Text>
              </Flex>
              <Input
                name="upload"
                type="file"
                hidden
                accept="image/png, image/webp, image/jpeg"
                onChange={handleUploadFile}
                ref={inputRef}
                data-testid="upload"
              />
            </Flex>
          )}
        </Flex>
      ) : (
        <Box
          borderRadius="md"
          border="1px solid"
          borderColor="borderCatSkill"
          px={2}
          pos="relative"
          w={width}
          h={height}
        >
          <Flex alignItems="center" justifyContent="center" h={boxHeightImage}>
            {isLoading ? (
              <Flex justifyContent="center" alignItems="center">
                <Spinner label="Uploading..." />
              </Flex>
            ) : (
              <CustomImage
                alt={alt}
                src={previewImage}
                width="200"
                height="200"
                style={{ objectFit: 'contain' }}
                blurDataURL={BLUR_DATA_URL}
              />
            )}
          </Flex>
          <Flex pos="absolute" top={1} right={0}>
            <Box onClick={handleButtonClick}>
              <Input
                name="upload"
                type="file"
                hidden
                accept="image/png, image/webp, image/jpeg"
                onChange={handleUploadFile}
                ref={inputRef}
              />
              <IconButton
                aria-label="re-upload-icon"
                background="none"
                _hover="none"
                icon={<UploadIcon width="20" height="20" />}
              />
            </Box>
            <IconButton
              background="none"
              _hover="none"
              aria-label="delete-icon"
              icon={<TrashIcon stroke="black" width="20" height="20" />}
              onClick={handleRemoveImage}
            />
          </Flex>
        </Box>
      )}
    </>
  );
};

export default UploadImage;
