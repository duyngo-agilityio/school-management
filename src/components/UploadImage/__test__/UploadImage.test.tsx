import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

// Hooks
import UploadImage from '..';


const mockHandleUploadImage = jest.fn();

// Mock the useUpload hook
jest.mock('@/hooks', () => ({
  useUpload: jest.fn().mockImplementation(() => ({
    isLoading: false,
    handleUploadImage: mockHandleUploadImage,
  })),
}));

describe('UploadImage Component', () => {
  const defaultProps = {
    title: 'Upload Image',
    height: '200px',
    width: '200px',
    alt: 'Uploaded Image',
    name: 'test-image',
    onChange: jest.fn(),
    onRemove: jest.fn(),
  };

  let container: ReturnType<typeof render>;

  beforeEach(() => {
    container = render(<UploadImage {...defaultProps} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders match snapshot', () => {
    expect(container.container).toMatchSnapshot();
  });

  it('handles image upload and calls onChange', async () => {
    const file = new File(['dummy content'], 'example.png', {
      type: 'image/png',
    });

    const dataImage = {
      image: { name: 'example.png', url: 'http://example.com/image.png' },
    };

    mockHandleUploadImage.mockResolvedValue(dataImage);

    const uploadButton = container.getByText('Upload Image');
    fireEvent.click(uploadButton);

    const input = container.getByTestId('upload') as HTMLInputElement;
    fireEvent.change(input, { target: { files: [file] } });

    expect(mockHandleUploadImage).toHaveBeenCalledWith(file);
  });

  it('handles image removal and calls onRemove', () => {
    const propsWithImage = {
      ...defaultProps,
      src: 'http://example.com/image.png',
    };

    render(<UploadImage {...propsWithImage} />);

    const deleteButton = screen.getByLabelText('delete-icon');
    fireEvent.click(deleteButton);

    expect(defaultProps.onRemove).toHaveBeenCalledWith('test-image');
    expect(screen.queryByAltText('Uploaded Image')).not.toBeInTheDocument();
  });

  it('displays a toast on upload failure', async () => {
    const file = new File(['dummy content'], 'example.png', {
      type: 'image/png',
    });
    mockHandleUploadImage.mockResolvedValue({
      error: {
        name: 'upload error',
        message: 'Upload image fail',
      },
    });

    const uploadButton = container.getByText('Upload Image');
    fireEvent.click(uploadButton);

    const input = container.getByTestId('upload') as HTMLInputElement;
    fireEvent.change(input, { target: { files: [file] } });

    waitFor(() => {
      expect(container.getByText('Upload image fail')).toBeInTheDocument();
    });
  });
});
