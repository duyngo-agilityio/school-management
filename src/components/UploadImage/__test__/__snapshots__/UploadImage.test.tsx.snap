// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadImage Component renders match snapshot 1`] = `
<div>
  <div
    class="css-1ya6kzr"
  >
    <div
      class="css-1jij6rs"
    >
      <div
        class="css-759u60"
      >
        <svg
          fill="none"
          height="24px"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 4.5L12 14.5M12 4.5C11.2998 4.5 9.99153 6.4943 9.5 7M12 4.5C12.7002 4.5 14.0085 6.4943 14.5 7"
            stroke="#141B34"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M20 16.5C20 18.982 19.482 19.5 17 19.5H7C4.518 19.5 4 18.982 4 16.5"
            stroke="#141B34"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
        <p
          class="chakra-text css-1yhdc23"
        >
          Upload Image
        </p>
      </div>
      <input
        accept="image/png, image/webp, image/jpeg"
        class="chakra-input css-0"
        data-testid="upload"
        hidden=""
        name="upload"
        type="file"
      />
    </div>
  </div>
</div>
`;
